include/deal.II/base/thread_local_storage.h

```diff
@@ -193,40 +184,75 @@ namespace Threads
   // ----------------- inline and template functions --------------------------
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
-    : data(t)
+  inline ThreadLocalStorage<T>::ThreadLocalStorage()
   {}
 
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(
-    const ThreadLocalStorage<T> &t)
-    : data(t)
-  {}
+  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
+  {
+    exemplar = std::make_shared<T>(t);
+  }
 
 
   template <typename T>
   inline T &
-  ThreadLocalStorage<T>::get()
+  ThreadLocalStorage<T>::get(bool &exists)
   {
-#    ifdef DEAL_II_WITH_THREADS
-    return data.local();
-#    else
-    return data;
-#    endif
+    const std::thread::id my_id = std::this_thread::get_id();
+
+    // Note that std::map<..>::emplace guarantees that no iterators or
+    // references to stored objects are invalidated. We thus only have to
+    // ensure that we do not performa a lookup while writing, and that we
+    // do not write concurrently. This is precisely the "reader-writer
+    // lock" paradigm supported by C++14 by means of the std::shared_lock
+    // and th std::unique_lock
+
+    {
+      // Take a shared ("reader") lock for lookup and record the fact
+      // whether we could find an entry in the boolean exists.
+      std::shared_lock<decltype(insertion_mutex)> lock(insertion_mutex);
+      const auto                                  it = data.find(my_id);
+      if (it != data.end())
+        {
+          exists = true;
+          return it->second;
+        }
+      else
+        {
+          exists = false;
+        }
+    }
+
+    {
+      // Take a unique ("writer") lock for manipulating the std::map. This
+      // lock ensures that no other thread does a lookup at the same time.
+      //
+      std::unique_lock<decltype(insertion_mutex)> lock(insertion_mutex);
+
+      if constexpr (std::is_trivially_copyable<T>::value)
+        if (exemplar)
+          {
+            const auto it =
+              data.emplace(std::make_pair(my_id, *exemplar)).first;
```

Comment:
You can save one move operation using
```suggestion
              data.emplace(my_id, *exemplar).first;
```